dvc/ignore.py

```diff
@@ -176,60 +115,65 @@ def is_ignored_file(self, _):
 
 class DvcIgnoreFilter:
     def __init__(self, tree, root_dir):
+        from dvc.repo import Repo
+
         self.tree = tree
         self.root_dir = root_dir
-        self.ignores = {
-            DvcIgnoreDirs([".git", ".hg", ".dvc"]),
-            DvcIgnoreRepo(),
-        }
-        ignore_pattern_trie = DvcIgnorePatternsTrie()
+        default_ignore_pattern = [".hg/", ".git/", "{}/".format(Repo.DVC_DIR)]
+        self.ignores_trie_tree = StringTrie(separator=os.sep)
+        self.ignores_trie_tree[root_dir] = DvcIgnorePatterns(
+            default_ignore_pattern, root_dir
+        )
         for root, dirs, _ in self.tree.walk(self.root_dir):
-            ignore_pattern = self._get_ignore_pattern(root)
-            if ignore_pattern:
-                ignore_pattern_trie[root] = ignore_pattern
-                self.ignores.add(ignore_pattern_trie)
+            self._update(root)
             dirs[:], _ = self(root, dirs, [])
 
-    def _get_ignore_pattern(self, dirname):
+    def _update(self, dirname):
         ignore_file_path = os.path.join(dirname, DvcIgnore.DVCIGNORE_FILE)
         if self.tree.exists(ignore_file_path):
-            return DvcIgnorePatterns.from_files(ignore_file_path, self.tree)
-        return None
+            new_pattern = DvcIgnorePatterns.from_files(
+                ignore_file_path, self.tree
+            )
+            old_pattern = self._get_tire_pattern(dirname)
+            self.ignores_trie_tree[dirname] = DvcIgnorePatterns(
+                *merge_patterns(
+                    old_pattern.pattern_list,
+                    old_pattern.dirname,
+                    new_pattern.pattern_list,
+                    new_pattern.dirname,
+                )
+            )
 
     def __call__(self, root, dirs, files):
-        for ignore in self.ignores:
-            dirs, files = ignore(root, dirs, files)
+        ignore_pattern = self._get_tire_pattern(root)
+        return ignore_pattern(root, dirs, files)
 
-        return dirs, files
+    def _get_tire_pattern(self, dirname):
+        ignore_pattern = self.ignores_trie_tree.longest_prefix(dirname).value
+        if not ignore_pattern:
+            return DvcIgnorePatterns([], dirname)
```

Comment:
Shouldn't we just return None here and react properly?
Now we will keep empty `DvcIgnorePatterns` for every directory in a tree, that does not containe `.dvcignore`. I think in big project, that might be causing some memory issues. And we already have [problems with that](https://github.com/iterative/dvc/issues/4139).

Reply:
@pared 
Thank you,  I know it's not a good design, but can't find an elegant solution.

We are now using a trie tree, and add a default pattern to the root of the Repo. So every `dirname` inside the repo would not get an empty pattern, they at least would return the pattern of the Repo's root. 

But if we call with some `dirname` outside the Repo, then it would return a None. Although no code would do that now(`_outside_repo` prevents it), I still add this logic to make it more robust. And I think an exception handling logic inside `_get_tire_pattern` is better than handling it in every use case.

The best solution I thought is to add something like `DVCIgnorePatternNoop` to the root of the fs.  And it could be done in `Unix` like system where we have `/`, but the `Windows` uses a different logic, there is no root of `C:\`, `D:\` etc.

The memory usage would not be a question in this situation, because the trie tree only had nodes where a `.dvcignore` file exists. Other directories in a tree would fall to these nodes.

Reply:
@karajan1001 
Thanks for the explanation, I get it now. 
I would still say that more readable would be to just react on `None`, due to understanding that one needs to grasp during parsing the code. But that seems more like a personal opinion, so let's not dwell into it.

Reply:
@pared 
As `_get_tire_pattern` is a private function, I think it is OK to grasp `None` outside it, as the callings are all under control.